flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/TimeBoundedStreamInnerJoin.scala

```diff
@@ -0,0 +1,533 @@
+    val rightStateDescriptor: MapStateDescriptor[Long, JList[Row]] =
+      new MapStateDescriptor[Long, JList[Row]](timeIndicator + "InnerJoinRightCache",
+        BasicTypeInfo.LONG_TYPE_INFO.asInstanceOf[TypeInformation[Long]], rightListTypeInfo)
+    rightCache = getRuntimeContext.getMapState(rightStateDescriptor)
+
+    // Initialize the timer states.
+    val leftTimerStateDesc: ValueStateDescriptor[Long] =
+      new ValueStateDescriptor[Long](timeIndicator + "InnerJoinLeftTimerState",
+        classOf[Long])
+    leftTimerState = getRuntimeContext.getState(leftTimerStateDesc)
+
+    val rightTimerStateDesc: ValueStateDescriptor[Long] =
+      new ValueStateDescriptor[Long](timeIndicator + "InnerJoinRightTimerState",
+        classOf[Long])
+    rightTimerState = getRuntimeContext.getState(rightTimerStateDesc)
+  }
+
+  /**
+    * Process records from the left stream.
+    *
+    * @param cRowValue the input record
+    * @param ctx       the context to register timer or get current time
+    * @param out       the collector for outputting results
+    *
+    */
+  override def processElement1(
+    cRowValue: CRow,
+    ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+    out: Collector[CRow]): Unit = {
+    val timeForRecord: Long = getTimeForRecord(ctx, cRowValue, true)
+    getCurrentOperatorTime(ctx)
+    processElement(
+      cRowValue,
+      timeForRecord,
+      ctx,
+      out,
+      leftOperatorTime,
+      rightOperatorTime,
+      rightTimerState,
+      leftCache,
+      rightCache,
+      true
+    )
+  }
+
+  /**
+    * Process records from the right stream.
+    *
+    * @param cRowValue the input record
+    * @param ctx       the context to get current time
+    * @param out       the collector for outputting results
+    *
+    */
+  override def processElement2(
+    cRowValue: CRow,
+    ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+    out: Collector[CRow]): Unit = {
+    val timeForRecord: Long = getTimeForRecord(ctx, cRowValue, false)
+    getCurrentOperatorTime(ctx)
+    processElement(
+      cRowValue,
+      timeForRecord,
+      ctx,
+      out,
+      rightOperatorTime,
+      leftOperatorTime,
+      leftTimerState,
+      rightCache,
+      leftCache,
+      false
+    )
+  }
+
+  /**
+    * Put a record from the input stream into the cache and iterate the opposite cache to
+    * output records meeting the join conditions. If there is no timer set for the OPPOSITE
+    * STREAM, register one.
+    */
+  private def processElement(
+    cRowValue: CRow,
+    timeForRecord: Long,
+    ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+    out: Collector[CRow],
+    myWatermark: Long,
+    oppositeWatermark: Long,
+    oppositeTimeState: ValueState[Long],
+    recordListCache: MapState[Long, JList[Row]],
+    oppositeCache: MapState[Long, JList[Row]],
+    leftRecord: Boolean): Unit = {
+    if (relativeWindowSize > 0) {
```

Comment:
If `relativeWindowSize <= 0`, the whole function will be executed without any problem, but no results can be produced. This check was added as an optimization. However, I think we should not reject this case (a warning is enough). After all, it's an improper condition instead of an error. What do you think?

Reply:
I see, but adding a condition at this place adds overhead for the common case that will (or should) always be true. 

I'd rather add a different operator that doesn't emit anything during plan construction.

Reply:
That's a good idea. I'll move this check out of the function.